File: LektaAndroid/app/src/main/java/com/lekta/app/viewmodels/ProductViewModel.kt
package com.lekta.app.viewmodels

import androidx.lifecycle.ViewModel
import com.lekta.app.models.Product
import com.lekta.app.services.SecurePrefs
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import javax.inject.Inject

@HiltViewModel
class ProductViewModel @Inject constructor(
    private val securePrefs: SecurePrefs
) : ViewModel() {

    private val _products = MutableStateFlow<List<Product>>(emptyList())
    val products: StateFlow<List<Product>> = _products

    private val storageKey = "wh_products_v2"

    init { load() }

    private fun load() {
        _products.value = securePrefs.loadEncrypted<List<Product>>(storageKey) ?: emptyList()
    }

    private fun save() {
        securePrefs.saveEncrypted(storageKey, _products.value)
    }

    fun find(barcode: String): Product? = _products.value.find { it.barcode == barcode }

    fun upsert(product: Product) {
        val list = _products.value.toMutableList()
        val idx = list.indexOfFirst { it.id == product.id }
        if (idx >= 0) list[idx] = product else list.add(product)
        _products.value = list
        save()
    }

    fun delete(product: Product) {
        _products.value = _products.value.filter { it.id != product.id }
        save()
    }

    fun addStock(barcode: String, qty: Int) {
        if (qty <= 0) return
        val list = _products.value.toMutableList()
        val idx = list.indexOfFirst { it.barcode == barcode }
        if (idx < 0) return
        list[idx] = list[idx].copy(stock = (list[idx].stock + qty).coerceAtMost(99_999))
        _products.value = list
        save()
    }

    fun removeStock(barcode: String, qty: Int) {
        if (qty <= 0) return
        val list = _products.value.toMutableList()
        val idx = list.indexOfFirst { it.barcode == barcode }
        if (idx < 0 || list[idx].stock < qty) return
        list[idx] = list[idx].copy(stock = list[idx].stock - qty)
        _products.value = list
        save()
    }

    val categories: List<String>
        get() = _products.value.map { it.category }.distinct().sorted()

    val totalStockValue: Double
        get() = _products.value.sumOf { it.finalPrice * it.stock }
}

File: ViewModels/ProductViewModel.swift
import Foundation
import Combine

class ProductViewModel: ObservableObject {
    @Published var products: [Product] = []

    private let key = "wh_products"

    init() {
        load()
    }

    // MARK: - Persistence

    private func load() {
        guard let data = UserDefaults.standard.data(forKey: key),
              let decoded = try? JSONDecoder().decode([Product].self, from: data) else { return }
        products = decoded
    }

    private func save() {
        guard let data = try? JSONEncoder().encode(products) else { return }
        UserDefaults.standard.set(data, forKey: key)
    }

    // MARK: - CRUD

    func find(barcode: String) -> Product? {
        products.first { $0.barcode == barcode }
    }

    func upsert(_ product: Product) {
        if let idx = products.firstIndex(where: { $0.id == product.id }) {
            products[idx] = product
        } else {
            products.append(product)
        }
        save()
    }

    func delete(_ product: Product) {
        products.removeAll { $0.id == product.id }
        save()
    }

    func addStock(barcode: String, qty: Int) {
        guard let idx = products.firstIndex(where: { $0.barcode == barcode }) else { return }
        products[idx].stock += qty
        save()
    }

    func removeStock(barcode: String, qty: Int) {
        guard let idx = products.firstIndex(where: { $0.barcode == barcode }),
              products[idx].stock >= qty else { return }
        products[idx].stock -= qty
        save()
    }

    // MARK: - Helpers

    var categories: [String] {
        Array(Set(products.map(\.category))).sorted()
    }

    var totalStockValue: Double {
        products.reduce(0) { $0 + $1.finalPrice * Double($1.stock) }
    }
}
